scripts/validate_string_concatenation.py

```diff
@@ -0,0 +1,133 @@
+#!/usr/bin/env python
+"""
+GH #30454
+
+Check where there is a string that needs to be concatenated.
+
+This is necessary after black formating,
+where for example black transforms this:
+
+>>> foo = (
+...         "bar "
+...         "baz"
+...     )
+
+into this:
+
+>>> foo = ("bar " "baz")
+
+Black is not considering this as an
+issue (see issue https://github.com/psf/black/issues/1051),
+so we are checking it here.
+"""
+
+import argparse
+import os
+import sys
+import token
+import tokenize
+from typing import Dict, Generator, List, Tuple
+
+FILE_EXTENSIONS_TO_CHECK: Tuple[str, str, str, str] = (
+    ".py",
+    ".pyx",
+    ".pyx.ini",
+    ".pxd",
+)
+
+
+def main(*, source_path: str = ".", output_format: str) -> bool:
+    """
+    Main entry point of the script.
+
+    Parameters
+    ----------
+    source_path : str, default '.'
+        Source path representing path to a file/directory.
+    output_format : str
+        Output format of the script.
+
+    Returns
+    -------
+    bool
+        True if found any strings that needs to be concatenated.
+
+    Raises
+    ------
+    ValueError
+        If the `source_path` is not pointing to existing file/directory.
+    """
+    if not os.path.exists(source_path):
+        raise ValueError(
+            "Please enter a valid path, pointing to a valid file/directory."
+        )
+
+    is_failed: bool = False
+
+    if os.path.isfile(source_path):
+        for values in strings_to_concatenate(source_path):
+            is_failed = True
+            print(output_format.format(**values))
+
+    for subdir, _, files in os.walk(source_path):
+        for file_name in files:
+            if any(
+                file_name.endswith(extension) for extension in FILE_EXTENSIONS_TO_CHECK
+            ):
+                for values in strings_to_concatenate(os.path.join(subdir, file_name)):
+                    is_failed = True
+                    print(output_format.format(**values))
```

Comment:
The output of this will be something like:
```
pandas/__init__.py:124:    BETWEEN first part of the string AND with a second part in another string
```

I don't think people will understand what's going on when seeing this error in the CI. I think we should display something like:
```
pandas/__init__.py:124:String unnecessarily split in two by black. Please merge them manually.
```

Reply:
I'm not sure about this, will the message be too long?
and where do I put the ```source_path, line_number, start, end```?

Reply:
The message being long is not a problem. Neither for the CI or for the code.

From my example, the format would be:
```
{source_path}:{line_number}:String unnecessarily split in two by black. Please merge them manually.
```

As I mentioned eariler, I don't find `start` and `end` particularly useful for the reader of the error. Feel free to leave them if you disagree, but better rename them, their name is misleading. `string1` and `string2` are not great names, but they're better. At least they don't create the wrong impression those are the positions of the strings.

Reply:
I can now see what you mean, Fixed.

Reply:
Cool, looks much better.

Sorry I said it wrong, but the format should actually be `{source_path}:{line_number}:{msg}`. We'll decide the error here. Imagine that this function can detect more than one error, we can't say in the format which is the error we want to receive.

So this could become:
```suggestion
                    msg = "String unnecessarily split in two by black. Please merge them manually."
                    print(output_format.format(source_path=source_path, line_number=line_number, msg=msg))
```

And since now the `strings_to_concatenate` function is just returning two values, I think it makes more sense to return a tuple with them, instead of a dict, so the `for above` would be:
```
for source_path, line_number in strings_to_concatenate(os.path.join(subdir, file_name)):
```